## 3D Data Representations
**1) 3D Point Cloud** - a collection of 3D points, where each point is represented by one 3-dimensional tuple (x, y, z).
Issues:
- From the DL point of view, 3D point clouds are one of the unordered and irregular data types: there is no clear and regular definition of the neighbourhood, so convolutions usually cannot be applied to point clouds. Thus, special types of DL models need to be used, like [PointNet](https://arxiv.org/abs/1612.00593). 
- Heterogeneous data issue, meaning for one training dataset, different point clouds may contain a different number of 3D points. Heterogeneous data can create some difficulties when using mini-batch Gradient Descent training.

**2) Mesh Representation** - another widely used 3D data representation. Each mesh contains a set of 3D points(vertices) and a set of polygons(faces), which are defined on the vertices. In most cases, meshes are obtained from post-processing from raw measurements of depth cameras or manually created during 3D asset design. 
Meshes contain additional geometric information and encode topology, and, like 3D point clouds, have heterogeneous data issues.

**3) Voxel Representation** - a representation that is based on voxels, a counterpart of pixels in 3D Computer Vision. A voxel is defined by dividing a 3D cube into small-sized cubes - each of the cubes is called a voxel. Important definitions related to voxels:
- SDF(Signed Distance Function): signed distance between the voxel center and the closest point on the surface (a positive sign means the voxel center is outside an object),
- TSDF(Truncated Signed Distance Function): truncated values of SDF, so they range from -1 to +1. 
Voxel representation is ordered and regular, so it is possible to use convolution filters. The potential disadvantage - voxel representation requires more computer memory.

## Working with Ply Files

In [1]:
# open3d package is handy for visualizing 3D data
# !pip install open3d
import open3d

from pytorch3d.io import load_ply 

In [2]:
data_folder = "data/"

### Example 1 - 'cube.ply'

In [3]:
# open3d package is handy for visualizing 3D data
# !pip install open3d
import open3d

from pytorch3d.io import load_ply 

In [4]:
mesh_file = data_folder+"cube.ply"

In [5]:
# Load the mesh 
mesh = open3d.io.read_triangle_mesh(mesh_file)

# Create a visualizer object and add the mesh
vis = open3d.visualization.Visualizer()
vis.create_window()
vis.add_geometry(mesh)

# Set mesh properties
mesh_show_wireframe = True
mesh_show_back_face = True

# Render the visualization and wait for window close
vis.get_render_option().mesh_show_wireframe = mesh_show_wireframe
vis.get_render_option().mesh_show_back_face = mesh_show_back_face
vis.run()
vis.destroy_window()

[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display
[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display


In [6]:
# Loading the same file with PyTorch3D
vertices, faces = load_ply(mesh_file) #

print('Type of vertices = ', type(vertices), ", type of faces = ", type(faces))
print('vertices = ', vertices)
print('faces = ', faces)

Type of vertices =  <class 'torch.Tensor'> , type of faces =  <class 'torch.Tensor'>
vertices =  tensor([[-1., -1., -1.],
        [ 1., -1., -1.],
        [ 1.,  1., -1.],
        [-1.,  1., -1.],
        [-1., -1.,  1.],
        [ 1., -1.,  1.],
        [ 1.,  1.,  1.],
        [-1.,  1.,  1.]])
faces =  tensor([[0, 1, 2],
        [5, 4, 7],
        [6, 2, 1],
        [3, 7, 4],
        [7, 3, 2],
        [5, 1, 0],
        [0, 2, 3],
        [5, 7, 6],
        [6, 1, 5],
        [3, 4, 0],
        [7, 2, 6],
        [5, 0, 4]])


### Example 2 - 'parallel_plane_mono.ply'

In [7]:
mesh_file = data_folder+"parallel_plane_mono.ply"

In [8]:
# Visualizing the mesh using open3D
mesh = open3d.io.read_triangle_mesh(mesh_file)
open3d.visualization.draw_geometries([mesh],
                                     mesh_show_wireframe = True,
                                     mesh_show_back_face = True,
                                     )

In [9]:
# Loading the same file with PyTorch3D
vertices, faces = load_ply(mesh_file)
print('Type of vertices = ', type(vertices), ", type of faces = ", type(faces))
print('vertices = ', vertices)
print('faces = ', faces)

Type of vertices =  <class 'torch.Tensor'> , type of faces =  <class 'torch.Tensor'>
vertices =  tensor([[-1., -1., -1.],
        [ 1., -1., -1.],
        [ 1.,  1., -1.],
        [-1.,  1., -1.],
        [-1., -1.,  1.],
        [ 1., -1.,  1.],
        [ 1.,  1.,  1.],
        [-1.,  1.,  1.]])
faces =  tensor([[0, 1, 2],
        [0, 2, 3],
        [5, 4, 7],
        [5, 7, 6]])


### Example 3 - 'parallel_plane_color.ply'

In the .ply file in this example, additional properties are defined for each vertex - red, green and blue properties of 'uchar' data type (see 'parallel_plane_color.ply' for details)

In [10]:
mesh_file = data_folder+"parallel_plane_color.ply"

In [11]:
# Visualizing the mesh using open3D
mesh = open3d.io.read_triangle_mesh(mesh_file)
open3d.visualization.draw_geometries([mesh],
                                     mesh_show_wireframe = True,
                                     mesh_show_back_face = True,
                                     )

[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display


In [12]:
# Loading the same file with PyTorch3D
vertices, faces = load_ply(mesh_file)
print('Type of vertices = ', type(vertices), ", type of faces = ", type(faces))
print('vertices = ', vertices)
print('faces = ', faces)

Type of vertices =  <class 'torch.Tensor'> , type of faces =  <class 'torch.Tensor'>
vertices =  tensor([[-1., -1., -1.],
        [ 1., -1., -1.],
        [ 1.,  1., -1.],
        [-1.,  1., -1.],
        [-1., -1.,  1.],
        [ 1., -1.,  1.],
        [ 1.,  1.,  1.],
        [-1.,  1.,  1.]])
faces =  tensor([[0, 1, 2],
        [0, 2, 3],
        [5, 4, 7],
        [5, 7, 6]])


### Example 4 - 'cow.ply'

In [13]:
mesh_file = data_folder+"cow.ply"

In [14]:
# Visualizing the mesh using open3D
mesh = open3d.io.read_triangle_mesh(mesh_file)
open3d.visualization.draw_geometries([mesh],
                                     mesh_show_wireframe = True,
                                     mesh_show_back_face = True,
                                     )

[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display


## Working with OBJ Files

### Example 1 - 'cube.obj'

In [15]:
from pytorch3d.io import load_obj

obj_file = data_folder+"cube.obj"

We will have a look at 'cube.obj' file. The first line in it, declares the companion MaterialTemplateLibrary(MTL) file - cube.mtl. 

The MTL file describes surface shading properties: 

- Ka: Specifies an ambient color
- Kd: Specifies a diffuse color
- Ks: Specifies a specular color
- Ns: Defines the focus of specular highlights
- Ni: Defines the optical density (a.k.a index of refraction)
- d: Specifies a factor for dissolve
- illum: Specifies an illumination model
- map_Kd: Specifies a color texture file to be applied to the diffuse reflectivity of the material

The 'cube.obj' file can be opened by both libraries - Open3D and PyTorch3D. 

In [16]:
# Visualizing the mesh using open3D
mesh = open3d.io.read_triangle_mesh(obj_file)
open3d.visualization.draw_geometries([mesh],
                 mesh_show_wireframe = True,
                 mesh_show_back_face = True)

[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display


In [17]:
# Loading the same file with PyTorch3D
vertices, faces, aux = load_obj(obj_file)

print('Type of vertices = ', type(vertices))
print("Type of faces = ", type(faces))
print("Type of aux = ", type(aux))
print('vertices = ', vertices)
print('faces = ', faces)
print('aux = ', aux)

Type of vertices =  <class 'torch.Tensor'>
Type of faces =  <class 'pytorch3d.io.obj_io.Faces'>
Type of aux =  <class 'pytorch3d.io.obj_io.Properties'>
vertices =  tensor([[-0.5000, -0.5000,  0.5000],
        [-0.5000, -0.5000, -0.5000],
        [-0.5000,  0.5000, -0.5000],
        [-0.5000,  0.5000,  0.5000],
        [ 0.5000, -0.5000,  0.5000],
        [ 0.5000, -0.5000, -0.5000],
        [ 0.5000,  0.5000, -0.5000],
        [ 0.5000,  0.5000,  0.5000]])
faces =  Faces(verts_idx=tensor([[0, 1, 2],
        [5, 4, 7],
        [6, 2, 1],
        [3, 7, 4],
        [7, 3, 2],
        [5, 1, 0],
        [0, 2, 3],
        [5, 7, 6],
        [6, 1, 5],
        [3, 4, 0],
        [7, 2, 6],
        [5, 0, 4]]), normals_idx=tensor([[-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1]]), textur

The returned **vertices** variable is a PyTorch tensor with a shape of 8x3. Each row is a vertex with the x, y, and z coordinates. 
The returned **faces** variable is a named tuple of three PyTorch tensors: verts_idx, normals_idx, and textures_idx. In the example, all the normals_idx and textures_idx tensors are invalid because the file "cube.obj" does not include definitions for normal and textures.

### Example 2 - 'cube_texture.obj'

The 'cube_texture.obj' file is similar to the file 'cube.obj' file, except for:
- Additional lines starting with vt: each such line declares a texture vertex with x and y coordinates. Each texture vertex defines a color. The color is the pixel color at a so-called texture image, where the pixel location is the x coordinate of the texture vertex x width, and the y coordinate of the texture vertex x height. The texture image would be defined in the cube_texture.mtl companion.
- There are additional lines starting with vn. Each such line declares a normal vector.
- Each face definition line now contains more information about each vertex.

The companion file is "cube_texture.mtl".
Again, it is possible to use  both Open3D and PyTorch3D to load the mesh in the "cube_texture.obj" file.

In [18]:
import open3d
from pytorch3d.io import load_obj
import torch

In [19]:
mesh_file = data_folder+"cube_texture.obj"

In [20]:
# Visualizing the mesh using open3D
mesh = open3d.io.read_triangle_mesh(mesh_file)
open3d.visualization.draw_geometries([mesh],
                  mesh_show_wireframe = True,
                  mesh_show_back_face = True)

[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display


In [21]:
# Loading the same file with PyTorch3D
vertices, faces, aux = load_obj(mesh_file)

print('Type of vertices = ', type(vertices))
print("Type of faces = ", type(faces))
print("Type of aux = ", type(aux))

print('vertices = ', vertices)
print('faces = ', faces)
print('aux = ', aux)

Type of vertices =  <class 'torch.Tensor'>
Type of faces =  <class 'pytorch3d.io.obj_io.Faces'>
Type of aux =  <class 'pytorch3d.io.obj_io.Properties'>
vertices =  tensor([[ 1.0000, -1.0000, -1.0000],
        [ 1.0000, -1.0000,  1.0000],
        [-1.0000, -1.0000,  1.0000],
        [-1.0000, -1.0000, -1.0000],
        [ 1.0000,  1.0000, -1.0000],
        [ 1.0000,  1.0000,  1.0000],
        [-1.0000,  1.0000,  1.0000],
        [-1.0000,  1.0000, -1.0000]])
faces =  Faces(verts_idx=tensor([[1, 2, 3],
        [7, 6, 5],
        [4, 5, 1],
        [5, 6, 2],
        [2, 6, 7],
        [0, 3, 7],
        [0, 1, 3],
        [4, 7, 5],
        [0, 4, 1],
        [1, 5, 2],
        [3, 2, 7],
        [4, 0, 7]]), normals_idx=tensor([[0, 0, 0],
        [1, 1, 1],
        [2, 2, 2],
        [3, 3, 3],
        [4, 4, 4],
        [5, 5, 5],
        [0, 0, 0],
        [1, 1, 1],
        [2, 2, 2],
        [3, 3, 3],
        [4, 4, 4],
        [5, 5, 5]]), textures_idx=tensor([[ 0,  1,  2],
       

In [22]:
texture_images = getattr(aux, 'texture_images')
print('texture_images type = ', type(texture_images))
for key in texture_images:
    print(key)

print(texture_images['Skin'].shape)

texture_images type =  <class 'dict'>
Skin
torch.Size([250, 250, 3])


The fields 'normals_idx' and 'textures_idx' of 'faces' contain valid indices now (instead of taking a -1 value). The field 'normals' of 'aux' is a PyTorch tensor now (instead of being None). The field 'verts_uvs' of 'aux' is a PyTorch tensor now (instead of being None).

The field 'texture_images' of the 'aux' variable is not an empty dictionary any longer. The texture_images dictionary contains one entry with a key, Skin, and a PyTorch tensor with a shape of (250, 250, 3). This tensor is exactly the same as the image contained in the wal67ar_small.jpg file, as defined in the mtl_texture.mtl file.